### PR TITLE
Fixed handling of a list of coupling_maps passed to transpile()

### LINKED ISSUE #9885: Logical error in transpiler.py
### Environment

- **Qiskit Terra version**: 7ae670a8e74d5a98dd516f696530c74e95f73885
- **Python version**: n/a
- **Operating system**: n/a

### What is happening?

While debugging other issues with the coupling_map, I noticed a puzzling section in the code:

https://github.com/Qiskit/qiskit-terra/blob/1282a04d528dd56d8e401dccc388a32178e9c6cc/qiskit/compiler/transpiler.py#L358-L361

At this point, `coupling_map` is either a `CouplingMap`, a list of lists of integers, or `None` and `unique_transpile_args` is a list of dicts with these keys:

https://github.com/Qiskit/qiskit-terra/blob/1282a04d528dd56d8e401dccc388a32178e9c6cc/qiskit/compiler/transpiler.py#L757-L763

So I don't see how this condition could ever be true.

### How can we reproduce the issue?

Here is some code that can raise an exception because of the problem with the code:

```
from qiskit import QuantumCircuit, transpile

qc = QuantumCircuit(1)

transpile([qc] * 2, coupling_map=[None, None])
```

### What should happen?

No error for the above code.

### Any suggestions?

I think code needs to check if `"coupling_map"` is in the shared dict, pull the coupling maps out of the unique dicts if not.

### PR DESCRIPTION
**Pull request by wshanks**
Closes #9885

### CODE REVIEW

**Comment by wshanks** [comment]
I can't think of why you would really want to pass a list of different coupling maps to `transpile` and I see that doing so is already marked for deprecation. I think the main benefit of this PR is sparing someone else like me from staring at them trying to figure how they could work :slightly_smiling_face:

**Review by mtreinish — COMMENTED** [review]
Lol, yeah that could never work, that definitely slipped through on review in #7789. I guess it kind of confirms the argument broadcasting in `transpile()` is not widely used.  Since this is technically a bugfix it'd be good to have a test case to cover this (it's definitely **not** worth a release note). Do you mind adding a quick test that is using this deprecated functionality?

File: qiskit/compiler/transpiler.py [hunk 1]
```diff
@@ -355,8 +355,8 @@ def callback_func(**kwargs):
         ignore_backend_supplied_default_methods,
     )
     # Get transpile_args to configure the circuit transpilation job(s)
-    if coupling_map in unique_transpile_args:
-        cmap_conf = unique_transpile_args["coupling_map"]
+    if "coupling_map" not in shared_args:
+        cmap_conf = [a["coupling_map"] for a in unique_transpile_args]
```
Comment (wshanks) [inline qiskit/compiler/transpiler.py:359]:
```suggestion
        cmap_conf = [a["pass_manager_config"]["coupling_map"] for a in unique_transpile_args]
```

**Comment by wshanks** [comment]
Yes, adding a test is a good idea, especially since my first pass at the PR was broken :slightly_smiling_face:

**Comment by wshanks** [comment]
I added a test but it failed because it triggers the deprecation warning. How should one test code that triggers a deprecation warning?

**Review by mtreinish — APPROVED** [review]
LGTM, thanks for adding the test

**Pull request merged**